src/hotspot/share/services/diagnosticCommand.cpp

```diff
@@ -851,15 +851,21 @@ void CodeCacheDCmd::execute(DCmdSource source, TRAPS) {
 }
 
 #ifdef LINUX
+#define DEFAULT_PERFMAP_FILENAME "/tmp/perf-<pid>.map"
+
 PerfMapDCmd::PerfMapDCmd(outputStream* output, bool heap) :
              DCmdWithParser(output, heap),
-  _filename("filename", "Name of the map file", "STRING", false)
+  _filename("filename", "Name of the map file", "STRING", false, DEFAULT_PERFMAP_FILENAME)
 {
   _dcmdparser.add_dcmd_argument(&_filename);
 }
 
 void PerfMapDCmd::execute(DCmdSource source, TRAPS) {
-  CodeCache::write_perf_map(_filename.value());
+  const char* filename = _filename.value();
+  if (strncmp(filename, DEFAULT_PERFMAP_FILENAME, strlen(DEFAULT_PERFMAP_FILENAME)) == 0) {
+      filename = nullptr; // use the default filename based on the pid
```

Comment:
Nit: the indent has to be 2, not 4.

Reply:
I only expected to see a change to the help string, I don't understand why the actual Dcmd is being modified here! ???

Reply:
The default value for the argument is what gets displayed in the help text. For example:

```
ThreadDumpToFileDCmd::ThreadDumpToFileDCmd(outputStream* output, bool heap) :
                                           DCmdWithParser(output, heap),
  _overwrite("-overwrite", "May overwrite existing file", "BOOLEAN", false, "false"),
  _format("-format", "Output format (\"plain\" or \"json\")", "STRING", false, "plain"),
  _filepath("filepath", "The file path to the output file", "STRING", true) {
  _dcmdparser.add_dcmd_option(&_overwrite);
  _dcmdparser.add_dcmd_option(&_format);
  _dcmdparser.add_dcmd_argument(&_filepath);
}
```

And the help text:

```
Options: (options must be specified using the <key> or <key>=<value> syntax)
	-overwrite : [optional] May overwrite existing file (BOOLEAN, false)
	-format : [optional] Output format ("plain" or "json") (STRING, plain)
```

The help output that indicates that "plain" is the default format comes from the intialization of the _format argument. There is no separate help text.

Reply:
Ugghh! So the help text is an actual stringification of the actual default value of the "field", whereas in this case the real default value comes from passing null to `CodeCache::write_perf_map`. So we need this hack to deal with that. That is truly awful IMO. The only way to cleanly address that is to expand things so that you can set an actual help string to be used.

Reply:
This check is problematic:

```
if (strncmp(filename, DEFAULT_PERFMAP_FILENAME, strlen(DEFAULT_PERFMAP_FILENAME)) == 0)
```

Because it cannot tell whether `filename` was explicitly given by the user. As a result, if the user specifies:

```
jcmd 1234 perfmap '/tmp/perf-<pid>.map'
```

the file will be written as `/tmp/perf-1234.map`, but if the user specifies

```
jcmd 1234 perfmap '/tmp/<pid>-perf.map'
```

then the file will be written as  `/tmp/<pid>-perf.map`.

This gives the confusing impression that the string `<pid>` is sometimes substituted and sometimes not.

I think it's better to do this (similarly for the `VM.cds` case):

```
void PerfMapDCmd::execute(DCmdSource source, TRAPS) {
  CodeCache::write_perf_map(_filename.is_set() ? _filename.value() : nullptr);
}
```

This essentially makes the JVM behavior exactly the same as before, so I think we can change the JBS issue title to something like "Clarify docs for filename parameter to Compiler.perfmap and VM.cds".

Reply:
@iklam but IIUC that does not address the issue of updationg the help output.

Reply:
My suggestion is to be used in combination with this change:

```
_filename("filename", "Name of the map file", "STRING", false, DEFAULT_PERFMAP_FILENAME)
```

So the help message will be printed as: 

```
*filename*: (Optional) The name of the map file (STRING /tmp/perf-<pid>.map)
```

Reply:
But that only works if `init_value` is never called, else `is_set` will be true. This seems a rather fragile way to have the "default string" available for the help command, but avoiding actually using it as a default value.

That said, after having a much deeper look at how this all hangs together, it may be this will have to suffice, at least for now. Really we need a way to express that the default-string is only intended for providing the help information it is not actually intended to be parsed to "generate" the default value. But I will also note that part of the problem here is that the DCmd help needs to report the default filename, but the DCmd is not actually responsible for establishing what that default is - we have to hard-wire into the DCMd help knowledge of what the underlying API does.

Reply:
For the dcmd help output, we could take a dynamic approach and have the dcmd call an API (which does not currently exist) which is responsible for generating the default file name. For CDS, this would have to be an upcall to java since that is where the default filename is currently generated.

...or we could just punt on this for now and continue with not actually specifying the default in the help output or the man page.

Reply:
I don't think we want that for the help output - I think we want the generic form with the "<pid>" text, not the filename with the actual pid value inserted.

I do think we want the help and manpage to document the default filenames. So I think Ioi's suggestion is the best option at the moment. Thanks.